File: airbyte-metrics/reporter/src/test/java/io/airbyte/metrics/reporter/EmitterTest.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.metrics.reporter;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStatus;
import io.airbyte.metrics.lib.MetricAttribute;
import io.airbyte.metrics.lib.MetricClient;
import io.airbyte.metrics.lib.MetricTags;
import io.airbyte.metrics.lib.OssMetricsRegistry;
import io.airbyte.metrics.reporter.model.LongRunningJobMetadata;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

@SuppressWarnings({"MethodName", "PMD.JUnitTestsShouldIncludeAssert"})
class EmitterTest {

  private MetricClient client;
  private MetricRepository repo;

  private static final String SYNC_QUEUE = "SYNC";
  private static final String AWS_QUEUE = "AWS";

  private static final String EU_REGION = "EU";
  private static final String AUTO_REGION = "AUTO";

  @BeforeEach
  void setUp() {
    client = mock(MetricClient.class);
    repo = mock(MetricRepository.class);
  }

  @Test
  void TestNumPendingJobs() {
    final var value = Map.of(AUTO_REGION, 101, EU_REGION, 20);
    when(repo.numberOfPendingJobsByGeography()).thenReturn(value);

    final var emitter = new NumPendingJobs(client, repo);
    emitter.emit();

    assertEquals(Duration.ofSeconds(15), emitter.getDuration());
    verify(repo).numberOfPendingJobsByGeography();
    verify(client).gauge(OssMetricsRegistry.NUM_PENDING_JOBS, 101,
        new MetricAttribute(MetricTags.GEOGRAPHY, AUTO_REGION));
    verify(client).gauge(OssMetricsRegistry.NUM_PENDING_JOBS, 20,
        new MetricAttribute(MetricTags.GEOGRAPHY, EU_REGION));
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestNumRunningJobs() {
    final var value = Map.of(SYNC_QUEUE, 101, AWS_QUEUE, 20);
    when(repo.numberOfRunningJobsByTaskQueue()).thenReturn(value);

    final var emitter = new NumRunningJobs(client, repo);
    emitter.emit();

    assertEquals(Duration.ofSeconds(15), emitter.getDuration());
    verify(repo).numberOfRunningJobsByTaskQueue();
    verify(client).gauge(OssMetricsRegistry.NUM_RUNNING_JOBS, 101,
        new MetricAttribute(MetricTags.ATTEMPT_QUEUE, SYNC_QUEUE));
    verify(client).gauge(OssMetricsRegistry.NUM_RUNNING_JOBS, 20,
        new MetricAttribute(MetricTags.ATTEMPT_QUEUE, AWS_QUEUE));
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestNumOrphanRunningJobs() {
    final var value = 101;
    when(repo.numberOfOrphanRunningJobs()).thenReturn(value);

    final var emitter = new NumOrphanRunningJobs(client, repo);
    emitter.emit();

    assertEquals(Duration.ofSeconds(15), emitter.getDuration());
    verify(repo).numberOfOrphanRunningJobs();
    verify(client).gauge(OssMetricsRegistry.NUM_ORPHAN_RUNNING_JOBS, value);
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestOldestRunningJob() {
    final var value = Map.of(SYNC_QUEUE, 101.0, AWS_QUEUE, 20.0);
    when(repo.oldestRunningJobAgeSecsByTaskQueue()).thenReturn(value);

    final var emitter = new OldestRunningJob(client, repo);
    emitter.emit();

    assertEquals(Duration.ofSeconds(15), emitter.getDuration());
    verify(repo).oldestRunningJobAgeSecsByTaskQueue();
    verify(client).gauge(OssMetricsRegistry.OLDEST_RUNNING_JOB_AGE_SECS, 101,
        new MetricAttribute(MetricTags.ATTEMPT_QUEUE, SYNC_QUEUE));
    verify(client).gauge(OssMetricsRegistry.OLDEST_RUNNING_JOB_AGE_SECS, 20,
        new MetricAttribute(MetricTags.ATTEMPT_QUEUE, AWS_QUEUE));
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestOldestPendingJob() {
    final var value = Map.of(AUTO_REGION, 101.0, EU_REGION, 20.0);
    when(repo.oldestPendingJobAgeSecsByGeography()).thenReturn(value);

    final var emitter = new OldestPendingJob(client, repo);
    emitter.emit();

    assertEquals(Duration.ofSeconds(15), emitter.getDuration());
    verify(repo).oldestPendingJobAgeSecsByGeography();
    verify(client).gauge(OssMetricsRegistry.OLDEST_PENDING_JOB_AGE_SECS, 101,
        new MetricAttribute(MetricTags.GEOGRAPHY, AUTO_REGION));
    verify(client).gauge(OssMetricsRegistry.OLDEST_PENDING_JOB_AGE_SECS, 20,
        new MetricAttribute(MetricTags.GEOGRAPHY, EU_REGION));

    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestNumActiveConnectionsPerWorkspace() {
    final var values = List.of(101L, 202L);
    when(repo.numberOfActiveConnPerWorkspace()).thenReturn(values);

    final var emitter = new NumActiveConnectionsPerWorkspace(client, repo);
    emitter.emit();

    assertEquals(Duration.ofSeconds(15), emitter.getDuration());
    verify(repo).numberOfActiveConnPerWorkspace();
    for (final var value : values) {
      verify(client).distribution(OssMetricsRegistry.NUM_ACTIVE_CONN_PER_WORKSPACE, value);
    }
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestNumAbnormalScheduledSyncs() {
    final var value = 101;
    when(repo.numberOfJobsNotRunningOnScheduleInLastDay()).thenReturn((long) value);

    final var emitter = new NumAbnormalScheduledSyncs(client, repo);
    emitter.emit();

    assertEquals(Duration.ofHours(1), emitter.getDuration());
    verify(repo).numberOfJobsNotRunningOnScheduleInLastDay();
    verify(client).gauge(OssMetricsRegistry.NUM_ABNORMAL_SCHEDULED_SYNCS_IN_LAST_DAY, value);
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestTotalScheduledSyncs() {
    final var value = 101;
    when(repo.numScheduledActiveConnectionsInLastDay()).thenReturn((long) value);

    final var emitter = new TotalScheduledSyncs(client, repo);
    emitter.emit();

    assertEquals(Duration.ofHours(1), emitter.getDuration());
    verify(repo).numScheduledActiveConnectionsInLastDay();
    verify(client).gauge(OssMetricsRegistry.NUM_TOTAL_SCHEDULED_SYNCS_IN_LAST_DAY, value);
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void TestTotalJobRuntimeByTerminalState() {
    final var values = Map.of(JobStatus.cancelled, 101.0, JobStatus.succeeded, 202.0,
        JobStatus.failed, 303.0);
    when(repo.overallJobRuntimeForTerminalJobsInLastHour()).thenReturn(values);

    final var emitter = new TotalJobRuntimeByTerminalState(client, repo);
    emitter.emit();

    assertEquals(Duration.ofHours(1), emitter.getDuration());
    verify(repo).overallJobRuntimeForTerminalJobsInLastHour();
    values.forEach((jobStatus, time) -> {
      verify(client).distribution(
          OssMetricsRegistry.OVERALL_JOB_RUNTIME_IN_LAST_HOUR_BY_TERMINAL_STATE_SECS, time,
          new MetricAttribute(MetricTags.JOB_STATUS, jobStatus.getLiteral()));
    });
    verify(client).count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1);
  }

  @Test
  void unusuallyLongSyncs() {
    final var values = List.of(
        new LongRunningJobMetadata("sourceImg1", "destImg1", "connection1"),
        new LongRunningJobMetadata("sourceImg2", "destImg2", "connection2"),
        new LongRunningJobMetadata("sourceImg3", "destImg3", "connection3"));
    when(repo.unusuallyLongRunningJobs()).thenReturn(values);

    final var emitter = new UnusuallyLongSyncs(client, repo);
    emitter.emit();

    values.forEach(meta -> {
      verify(client).count(
          OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS, 1,
          new MetricAttribute(MetricTags.SOURCE_IMAGE, meta.sourceDockerImage()),
          new MetricAttribute(MetricTags.DESTINATION_IMAGE, meta.destinationDockerImage()),
          new MetricAttribute(MetricTags.CONNECTION_ID, meta.connectionId()));
    });
  }

  @Test
  void unusuallyLongSyncsHandlesNullMetadata() {
    final List<LongRunningJobMetadata> values = new ArrayList<>();
    values.add(new LongRunningJobMetadata("sourceImg1", "destImg1", "connection1"));
    values.add(null); // specifically add a null to simulate a mapping failure
    values.add(new LongRunningJobMetadata("sourceImg2", "destImg2", "connection2"));
    when(repo.unusuallyLongRunningJobs()).thenReturn(values);

    final var emitter = new UnusuallyLongSyncs(client, repo);
    emitter.emit();

    // metric is incremented for well-formed job metadata with attrs
    verify(client).count(
        OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS, 1,
        new MetricAttribute(MetricTags.SOURCE_IMAGE, "sourceImg1"),
        new MetricAttribute(MetricTags.DESTINATION_IMAGE, "destImg1"),
        new MetricAttribute(MetricTags.CONNECTION_ID, "connection1"));

    verify(client).count(
        OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS, 1,
        new MetricAttribute(MetricTags.SOURCE_IMAGE, "sourceImg2"),
        new MetricAttribute(MetricTags.DESTINATION_IMAGE, "destImg2"),
        new MetricAttribute(MetricTags.CONNECTION_ID, "connection2"));

    // metric is incremented without attrs for the null case
    verify(client, times(1)).count(OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS, 1, new MetricAttribute[0]);
  }

}

File: airbyte-metrics/reporter/src/test/kotlin/io/airbyte/metrics/reporter/EmitterTest.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.metrics.reporter

import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStatus
import io.airbyte.metrics.MetricAttribute
import io.airbyte.metrics.MetricClient
import io.airbyte.metrics.OssMetricsRegistry
import io.airbyte.metrics.lib.MetricTags
import io.airbyte.metrics.reporter.model.LongRunningJobMetadata
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.Duration
import java.util.function.Consumer

internal class EmitterTest {
  lateinit var client: MetricClient
  lateinit var repo: MetricRepository

  @BeforeEach
  fun setUp() {
    client = mockk(relaxed = true)
    repo = mockk()
  }

  @Test
  fun testNumPendingJobs() {
    val value = mapOf(AUTO_REGION to 101, EU_REGION to 20)
    every { repo.numberOfPendingJobsByDataplaneGroupName() } returns value

    val emitter = NumPendingJobs(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofSeconds(15), emitter.getDuration())
    verify { repo.numberOfPendingJobsByDataplaneGroupName() }
    verify {
      client.gauge(
        OssMetricsRegistry.NUM_PENDING_JOBS,
        101.0,
        MetricAttribute(MetricTags.GEOGRAPHY, AUTO_REGION),
      )
    }
    verify {
      client.gauge(
        OssMetricsRegistry.NUM_PENDING_JOBS,
        20.0,
        MetricAttribute(MetricTags.GEOGRAPHY, EU_REGION),
      )
    }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testNumRunningJobs() {
    val value = mapOf(SYNC_QUEUE to 101, AWS_QUEUE to 20)
    every { repo.numberOfRunningJobsByTaskQueue() } returns value

    val emitter = NumRunningJobs(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofSeconds(15), emitter.getDuration())
    verify { repo.numberOfRunningJobsByTaskQueue() }
    verify {
      client.gauge(
        OssMetricsRegistry.NUM_RUNNING_JOBS,
        101.0,
        MetricAttribute(MetricTags.ATTEMPT_QUEUE, SYNC_QUEUE),
      )
    }
    verify {
      client.gauge(
        OssMetricsRegistry.NUM_RUNNING_JOBS,
        20.0,
        MetricAttribute(MetricTags.ATTEMPT_QUEUE, AWS_QUEUE),
      )
    }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testNumOrphanRunningJobs() {
    val value = 101
    every { repo.numberOfOrphanRunningJobs() } returns value

    val emitter = NumOrphanRunningJobs(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofSeconds(15), emitter.getDuration())
    verify { repo.numberOfOrphanRunningJobs() }
    verify { client.gauge(OssMetricsRegistry.NUM_ORPHAN_RUNNING_JOBS, value.toDouble()) }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testOldestRunningJob() {
    val value = mapOf(SYNC_QUEUE to 101.0, AWS_QUEUE to 20.0)
    every { repo.oldestRunningJobAgeSecsByTaskQueue() } returns value

    val emitter = OldestRunningJob(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofSeconds(15), emitter.getDuration())
    verify { repo.oldestRunningJobAgeSecsByTaskQueue() }
    verify {
      client.gauge(
        OssMetricsRegistry.OLDEST_RUNNING_JOB_AGE_SECS,
        101.0,
        MetricAttribute(MetricTags.ATTEMPT_QUEUE, SYNC_QUEUE),
      )
    }
    verify {
      client.gauge(
        OssMetricsRegistry.OLDEST_RUNNING_JOB_AGE_SECS,
        20.0,
        MetricAttribute(MetricTags.ATTEMPT_QUEUE, AWS_QUEUE),
      )
    }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testOldestPendingJob() {
    val value = mapOf(AUTO_REGION to 101.0, EU_REGION to 20.0)
    every { repo.oldestPendingJobAgeSecsByDataplaneGroupName() } returns value

    val emitter = OldestPendingJob(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofSeconds(15), emitter.getDuration())
    verify { repo.oldestPendingJobAgeSecsByDataplaneGroupName() }
    verify {
      client.gauge(
        OssMetricsRegistry.OLDEST_PENDING_JOB_AGE_SECS,
        101.0,
        MetricAttribute(MetricTags.GEOGRAPHY, AUTO_REGION),
      )
    }
    verify {
      client.gauge(
        OssMetricsRegistry.OLDEST_PENDING_JOB_AGE_SECS,
        20.0,
        MetricAttribute(MetricTags.GEOGRAPHY, EU_REGION),
      )
    }

    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testNumActiveConnectionsPerWorkspace() {
    val values = listOf(101L, 202L)
    every { repo.numberOfActiveConnPerWorkspace() } returns values

    val emitter = NumActiveConnectionsPerWorkspace(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofSeconds(15), emitter.getDuration())
    verify { repo.numberOfActiveConnPerWorkspace() }
    for (value in values) {
      verify { client.distribution(OssMetricsRegistry.NUM_ACTIVE_CONN_PER_WORKSPACE, value.toDouble()) }
    }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testNumAbnormalScheduledSyncs() {
    val value = 101
    every { repo.numberOfJobsNotRunningOnScheduleInLastDay() } returns value.toLong()

    val emitter = NumAbnormalScheduledSyncs(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofHours(1), emitter.getDuration())
    verify { repo.numberOfJobsNotRunningOnScheduleInLastDay() }
    verify { client.gauge(OssMetricsRegistry.NUM_ABNORMAL_SCHEDULED_SYNCS_IN_LAST_DAY, value.toDouble()) }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testTotalScheduledSyncs() {
    val value = 101
    every { repo.numScheduledActiveConnectionsInLastDay() } returns value.toLong()

    val emitter = TotalScheduledSyncs(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofHours(1), emitter.getDuration())
    verify { repo.numScheduledActiveConnectionsInLastDay() }
    verify { client.gauge(OssMetricsRegistry.NUM_TOTAL_SCHEDULED_SYNCS_IN_LAST_DAY, value.toDouble()) }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun testTotalJobRuntimeByTerminalState() {
    val values =
      mapOf(
        JobStatus.cancelled to 101.0,
        JobStatus.succeeded to 202.0,
        JobStatus.failed to 303.0,
      )
    every { repo.overallJobRuntimeForTerminalJobsInLastHour() } returns values

    val emitter = TotalJobRuntimeByTerminalState(client, repo)
    emitter.emit()

    Assertions.assertEquals(Duration.ofHours(1), emitter.getDuration())
    verify { repo.overallJobRuntimeForTerminalJobsInLastHour() }
    values.forEach { (jobStatus: JobStatus, time: Double) ->
      verify {
        client.distribution(
          OssMetricsRegistry.OVERALL_JOB_RUNTIME_IN_LAST_HOUR_BY_TERMINAL_STATE_SECS,
          time,
          MetricAttribute(MetricTags.JOB_STATUS, jobStatus.literal),
        )
      }
    }
    verify { client.count(OssMetricsRegistry.EST_NUM_METRICS_EMITTED_BY_REPORTER, 1L) }
  }

  @Test
  fun unusuallyLongSyncs() {
    val values =
      listOf(
        LongRunningJobMetadata("sourceImg1", "destImg1", "workspace1", "connection1"),
        LongRunningJobMetadata("sourceImg2", "destImg2", "workspace2", "connection2"),
        LongRunningJobMetadata("sourceImg3", "destImg3", "workspace3", "connection3"),
      )
    every { repo.unusuallyLongRunningJobs() } returns values

    val emitter = UnusuallyLongSyncs(client, repo)
    emitter.emit()

    values.forEach(
      Consumer { meta: LongRunningJobMetadata ->
        verify {
          client.count(
            OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS,
            1L,
            MetricAttribute(MetricTags.SOURCE_IMAGE, meta.sourceDockerImage),
            MetricAttribute(MetricTags.DESTINATION_IMAGE, meta.destinationDockerImage),
            MetricAttribute(MetricTags.CONNECTION_ID, meta.connectionId),
            MetricAttribute(MetricTags.WORKSPACE_ID, meta.workspaceId),
          )
        }
      },
    )
  }

  @Test
  fun unusuallyLongSyncsHandlesNullMetadata() {
    val values: MutableList<LongRunningJobMetadata?> = ArrayList()
    values.add(LongRunningJobMetadata("sourceImg1", "destImg1", "workspace1", "connection1"))
    values.add(null) // specifically add a null to simulate a mapping failure
    values.add(LongRunningJobMetadata("sourceImg2", "destImg2", "workspace2", "connection2"))
    @Suppress("UNCHECKED_CAST")
    every { repo.unusuallyLongRunningJobs() } returns (values as List<LongRunningJobMetadata>)

    val emitter = UnusuallyLongSyncs(client, repo)
    emitter.emit()

    // metric is incremented for well-formed job metadata with attrs
    verify {
      client.count(
        OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS,
        1L,
        MetricAttribute(MetricTags.SOURCE_IMAGE, "sourceImg1"),
        MetricAttribute(MetricTags.DESTINATION_IMAGE, "destImg1"),
        MetricAttribute(MetricTags.CONNECTION_ID, "connection1"),
        MetricAttribute(MetricTags.WORKSPACE_ID, "workspace1"),
      )
    }

    verify {
      client.count(
        OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS,
        1L,
        MetricAttribute(MetricTags.SOURCE_IMAGE, "sourceImg2"),
        MetricAttribute(MetricTags.DESTINATION_IMAGE, "destImg2"),
        MetricAttribute(MetricTags.CONNECTION_ID, "connection2"),
        MetricAttribute(MetricTags.WORKSPACE_ID, "workspace2"),
      )
    }

    // metric is incremented without attrs for the null case
    verify(exactly = 1) { client.count(OssMetricsRegistry.NUM_UNUSUALLY_LONG_SYNCS, 1L) }
  }

  companion object {
    private const val SYNC_QUEUE = "SYNC"
    private const val AWS_QUEUE = "AWS"

    private const val EU_REGION = "EU"
    private const val AUTO_REGION = "AUTO"
  }
}
